torchci/lib/bot/greenBot.ts

```diff
@@ -0,0 +1,212 @@
+import { Probot, ProbotOctokit } from "probot";
+import { reactOnComment, addLabels } from "./botUtils";
+
+const openOnGreen = new RegExp("^s*@pytorchbot\\s+?mergeOnGreen.*");
+export const LAND_PENDING = "land-pending";
+const LAND_ATTEMPT = "landed-attempt";
+const LAND_FAILED = "land-failed";
+
+type OctokitData = {
+  octokit: InstanceType<typeof ProbotOctokit>;
+  owner: string;
+  repo: string;
+  issue_number: number;
+  comment: string;
+};
+
+export default function greenBot(app: Probot): void {
+  app.on("issue_comment.created", async (ctx) => {
+    const commentBody = ctx.payload.comment.body;
+    if (commentBody.match(openOnGreen)) {
+      if (!ctx.payload.issue.pull_request) {
+        // Issue, not pull request.
+        await reactOnComment(ctx, "confused");
+      } else {
+        await reactOnComment(ctx, "+1");
+        await addLabels(ctx, [LAND_PENDING]);
+      }
+    }
+  });
+
+  app.on(["pull_request.labeled"], async (ctx) => {
+    if (ctx.payload.label.name === LAND_PENDING) {
+      // Check pull request's head ref and check runs to see if it has all green and merge if it's good
+      // TODO: Add pagination if checks go above 100
+      const payloadData = getPayloadData(ctx);
+      const ref = ctx.payload.pull_request.head.sha;
+      const prNum = ctx.payload.pull_request.number;
+      const checks = await ctx.octokit.checks.listForRef({
+        ...payloadData,
+        ref,
+        per_page: 100,
+      });
+
+      const [successfulChecks, failedChecks] = categorizeChecks(
+        checks.data.check_runs
+      );
+      // If all jobs are successful or neutral, then try merge, else do nothing
+      if (successfulChecks.length === checks.data.check_runs.length) {
+        await tryLandPR({
+          octokit: ctx.octokit,
+          ...payloadData,
+          issue_number: prNum,
+          comment: "All checks passed. Attempting to merge.",
+        });
+      }
+      // If any failed, alert the author
+      else if (failedChecks.length > 0) {
+        await failPR({
+          octokit: ctx.octokit,
+          ...payloadData,
+          issue_number: prNum,
+          comment: "Failed to land PR due to red signal.",
+        });
+      }
+    }
+  });
+
+  app.on("check_run.completed", async (ctx) => {
```

Comment:
As discussed: can we do status instead of check_run? Not all statuses have an associated check run.

Alternatively workflow_job might be okay. It's GHA only, but at least it's consistent with all the other workflow_job based stuff in the repo (so we can use the JobUtils stuff to parse failures and such)

Reply:
I looked at the statuses of some of the commits on master and all of them say pending, for example https://api.github.com/repos/pytorch/pytorch/commits/eb69e8a3edb73606afae5b91fad2be1f7a8ceb1d/status . 

Not sure if I'm doing something wrong. Also if we use status we won't be able to have conditionals on allowing them to merge if certain jobs are passing right, which is something we want right @janeyx99 ? I think in the future we would want that but might be a bit early. We can definitely change it to workflow_job though.

Reply:
That's the combined status API, there is one that lists all the stauses (https://api.github.com/repos/pytorch/pytorch/commits/eb69e8a3edb73606afae5b91fad2be1f7a8ceb1d/statuses), but it seems to be returning nothing right now. So workflow_job it is I guess lol—the primary limitation there is that it only supports GHA but I guess that's okay.